src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FeatureFlagPass.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ChildDefinition;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\FeatureFlag\Debug\TraceableFeatureChecker;
+
+class FeatureFlagPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('feature_flag.feature_checker')) {
+            return;
+        }
+
+        $features = [];
+        foreach ($container->findTaggedServiceIds('feature_flag.feature') as $serviceId => $tags) {
+            $className = $this->getServiceClass($container, $serviceId);
+            $r = $container->getReflectionClass($className);
+
+            if (null === $r) {
+                throw new \RuntimeException(\sprintf('Invalid service "%s": class "%s" does not exist.', $serviceId, $className));
+            }
+
+            foreach ($tags as $tag) {
+                $featureName = ($tag['feature'] ?? '') ?: $className;
+                if (\array_key_exists($featureName, $features)) {
+                    throw new \RuntimeException(\sprintf('Feature "%s" already defined in the "feature_flag.provider.in_memory" provider.', $featureName));
+                }
+
+                $method = $tag['method'] ?? '__invoke';
+                if (!$r->hasMethod($method)) {
+                    throw new \RuntimeException(\sprintf('Invalid feature method "%s": method "%s::%s()" does not exist.', $serviceId, $r->getName(), $method));
+                }
+
+                $features[$featureName] = $container->setDefinition(
+                    '.feature_flag.feature',
+                    (new Definition(\Closure::class))
+                        ->setLazy(true)
+                        ->setFactory([\Closure::class, 'fromCallable'])
+                        ->setArguments([[new Reference($serviceId), $method]]),
+                );
+            }
+        }
+
+        $container->getDefinition('feature_flag.provider.in_memory')
+            ->setArgument('$features', $features)
+        ;
+
+        if (!$container->has('feature_flag.data_collector')) {
+            return;
+        }
+
+        foreach ($container->findTaggedServiceIds('feature_flag.feature_checker') as $serviceId => $tags) {
+            $container->register('debug.'.$serviceId, TraceableFeatureChecker::class)
```

Comment:
should `debug` be prefixed with `.` as well ?

Reply:
Why this service should be hidden? It could be useful to know that a `FeatureChecker` has been decorated. 

PS: It does not seem to be a common usage in other Symfony components.